specifications/state_sync/README.md

```diff
@@ -0,0 +1,310 @@
+# State Synchronizer specification
+
+## Abstract
+
+This document describes the spec of a Libra node's state synchronizer component.
+
+This document is organized as follows:
+
+1. Overview: high-level outline of the the state synchronizer design and
+functionality.
+2. Sync Protocols: specify technical requirements of various sync protocols 
+recognized in the Libra Network.
+3. Network Protocol: the State Synchronizer LibraNet protocol used to participate
+as a node in the Libra Network
+4. Abstracted Modules: external components that state synchronizer interfaces with
+
+
+## Overview
+
+State Synchronizer is a system component that is responsible for syncing ledger
+state among peers in the Libra Payment Network. Using state sync, a Libra node
+can discover and update to a more recent ledger state if available with the following
+general flow:
+1. A node sends a [`GetChunkRequest`](#getchunkrequest) to a remote peer
+2. If there is a more recent ledger state available on the remote
+peer, it sends back a corresponding [`GetChunkResponse`](#getchunkresponse) with data that helps
+the requester advance its local ledger state
+
+
+There are three types of supported sync protocols:
+1. [Validator sync](#validator-sync): In general, validators learn about recent
+ledger state through their participation in the consensus protocol(TODO link),
+but it is possible they still fall behind (e.g. node goes offline for a while, 
+network partition). In these cases, a lagging validator can rely on state
+synchronization to catch up to the more recent state. 
+
+2. [Full Node sync](#full-node-sync): As full nodes do not participate in consensus,
+they rely only on the state synchronizer to discover and sync to a more recent
+ledger state. 
+
+3. [Waypoint sync](#waypoint-sync): When a node starts up, it can specify a waypoint
+to sync its local state to - see section 3.4 in [this paper](https://github.com/libra/libra/blob/master/documentation/tech-papers/lbft-verification/lbft-verification.pdf) for more context and 
+motivation for waypoint sync, as this specification primariliy focuses on
+the technical requirements for the waypoint sync protocol. Waypoint sync 
+happens at node initialization and must be completed before the node participates
+in other state sync protocols. 
+
+A properly implemented State Synchronizer component supports each
+sync protocol as both the requester and the responder. See [Sync Protocols](#sync-protocols)
+for the technical requirements for each of these sync protocols.
```

Comment:
perhaps a note on the fact that a single state-sync runs for all of these 3 points when you're a validator (as opposed to one state-sync for validator network and one state-sync for FN network)